io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -328,27 +331,65 @@ schedule thread@Thread{
       tvar  <- execNewTVar nextVid
                            (Just $ "<<timeout-state " ++ show (unTimeoutId nextTmid) ++ ">>")
                            TimeoutPending
-      tvar' <- execNewTVar (succ nextVid)
-                           (Just $ "<<timeout " ++ show (unTimeoutId nextTmid) ++ ">>")
-                           False
       let expiry  = d `addTime` time
-          t       = Timeout tvar tvar' nextTmid
-          timers' = PSQ.insert nextTmid expiry (TimerVars tvar tvar') timers
+          t       = Timeout tvar nextTmid
+          timers' = PSQ.insert nextTmid expiry (TimerFancy tvar) timers
           thread' = thread { threadControl = ThreadControl (k t) ctl }
+      trace <- schedule thread' simstate { timers   = timers'
+                                         , nextVid  = succ nextVid
+                                         , nextTmid = succ nextTmid }
+      return (SimTrace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) trace)
+
+    NewRegisterDelay d k | d < 0 -> do
+      tvar <- execNewTVar nextVid
+                          (Just $ "<<timeout " ++ show (unTimeoutId nextTmid) ++ ">>")
+                          True
+      let expiry  = d `addTime` time
+          thread' = thread { threadControl = ThreadControl (k tvar) ctl }
+      trace <- schedule thread' simstate { nextVid  = succ nextVid
+                                         }
+      return (SimTrace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) $
+              SimTrace time tid tlbl (EventTimerCancelled nextTmid) $
```

Comment:
We have `EventTimerCreated` (for `newTimer`) and `EventThreadDelay` (for `threadDelay`), I think we should also have `EventRegisterDelayCreated` (or `EventRegisteredDelayCreated`) (for `registerDelay`).

Reply:
done